### PR TITLE
New Feature: Adding Liquidity Pool Indexing for Moxie

### PR DESCRIPTION
**Pull request by Hrishikesh-Thakkar**
## Feature Description
The scope of this PR is to index 2 Liquidity Pools which are both based on Uniswap V2 Contracts

1. [Uniswap](https://basescan.org/address/0x493ad7e1c509de7c89e1963fe9005ead49fdd19c) (Moxie-USDC)
2. [Aerodrome](https://basescan.org/address/0xc02Dc3a9b6eaD739C56CDEa716D8fEC4Ffe4C799) (Moxie-WETH)

## Data Points Being Indexed
1. Reserves of both tokens in the Liquidity Pool
2. Total Supply of LP Tokens
3. Individual Balances of LP Tokens for each user
4. In the case of Aerodrome index the [LP Staking ](https://basescan.org/address/0xea2c1c05c17eed888d0027f17ee23637dff66feb) as well.

## Solution
1. To capture the balances and total supply with minimal RPC calls all the transfers of the LP Token are tracked.  
    Event: `Transfer(indexed address,indexed address,uint256)`
2. To capture the Reserves in both LPs we track the following event
    Event: `Sync(uint256,uint256)`  & `  Sync(uint112,uint112)`
3. To capture the Staking Contract related transactions we are capturing
    Event: `Deposit(indexed address,indexed address,uint256)` & `Withdraw(indexed address,uint256)`

### CODE REVIEW

**Review by sharathkrml — APPROVED** [review]

File: moxie-liquidity-pool/schema.graphql [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  createdAt: BlockInfo!
+  updatedAt: BlockInfo!
+}
+
+#Combined User and Pool entity
+type UserPool @entity {
+  id: ID!
+  user: User!
+  pool: Pool!
+  unstakedLpAmount: BigInt!
```
Comment (0xsarvesh) [inline moxie-liquidity-pool/schema.graphql:31]: can you add a field for totalLPAmount

File: moxie-liquidity-pool/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  //Currently in both aerodrome and uniswap reserve1 is always moxie. Because addresses are sorted when deciding which is token0 and token1. 
+  //Might need to extend in the future
+  pool.nonMoxieReserve = reserve0
+  pool.moxieReserve = reserve1
+  pool.save()
+}
+
+export function handleTransferEvents(event: ethereum.Event, from: Address, to: Address, amount: BigInt): void {
+  let pool = getOrCreatePoolEntity(event, event.address.toHexString())
+  let sender = getOrCreateUserEntity(event, from.toHexString())
```
Comment (0xsarvesh) [inline moxie-liquidity-pool/src/utils.ts:65]: check for zero from & to address. No need to track for zero address lp balance

File: moxie-liquidity-pool/src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  let receiverUserPool = getOrCreateUserPoolEntity(event, receiver.id, pool.id)
+  if (from == Address.zero()) {
+    // Mint: increase total supply
+    pool.totalSupply = pool.totalSupply.plus(amount)
+  } else if (to == Address.zero()) {
+    // Burn: decrease total supply
+    pool.totalSupply = pool.totalSupply.minus(amount)
+  }
+
+  senderUserPool.unstakedLpAmount = senderUserPool.unstakedLpAmount.minus(amount)
```
Comment (0xsarvesh) [inline moxie-liquidity-pool/src/utils.ts:78]: Similar comment, don't track for zero address.

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]

File: moxie-liquidity-pool/schema.graphql (hunk 1 above)
Comment (Hrishikesh-Thakkar) [reply moxie-liquidity-pool/schema.graphql:31]: done

**Review by 0xsarvesh — APPROVED** [review]

**Pull request merged**